executor/parallel_NLJ.go

```diff
@@ -0,0 +1,318 @@
+	"github.com/pingcap/tidb/util/stringutil"
+	"go.uber.org/zap"
+)
+
+type result struct {
+	chk *chunk.Chunk
+	err error
+}
+
+type outerRow struct {
+	row      *chunk.Row
+	selected bool // if this row is selected by the outer side
+}
+
+// ParallelNestedLoopApplyExec is the executor for apply.
+type ParallelNestedLoopApplyExec struct {
+	baseExecutor
+
+	// outer-side fields
+	cursor        int
+	outerExec     Executor
+	outerFilter   expression.CNFExprs
+	outerList     *chunk.List
+	outerRowMutex sync.Mutex
+	outer         bool
+
+	// inner-side fields
+	// use slices since the inner side is paralleled
+	corCols       [][]*expression.CorrelatedColumn
+	innerFilter   []expression.CNFExprs
+	innerExecs    []Executor
+	innerList     []*chunk.List
+	innerChunk    []*chunk.Chunk
+	innerSelected [][]bool
+	innerIter     []chunk.Iterator
+	outerRow      []*chunk.Row
+	hasMatch      []bool
+	hasNull       []bool
+
+	// fields are used to do concurrency control
+	concurrency int
+	started     bool
+	freeChkCh   chan *chunk.Chunk
+	resultChkCh chan result
+	outerRowCh  chan outerRow
+	numWorkers  int32
+	exit        chan struct{}
+	wg          sync.WaitGroup
+
+	joiner     joiner          // it's stateless
+	memTracker *memory.Tracker // track memory usage.
+}
+
+// Open implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Open(ctx context.Context) error {
+	err := e.outerExec.Open(ctx)
+	if err != nil {
+		return err
+	}
+	e.memTracker = memory.NewTracker(e.id, -1)
+	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+
+	e.outerList = chunk.NewList(retTypes(e.outerExec), e.initCap, e.maxChunkSize)
+	e.outerList.GetMemTracker().SetLabel(stringutil.StringerStr("outerList"))
+	e.outerList.GetMemTracker().AttachTo(e.memTracker)
+
+	e.innerList = make([]*chunk.List, e.concurrency)
+	e.innerChunk = make([]*chunk.Chunk, e.concurrency)
+	e.innerSelected = make([][]bool, e.concurrency)
+	e.innerIter = make([]chunk.Iterator, e.concurrency)
+	e.outerRow = make([]*chunk.Row, e.concurrency)
+	e.hasMatch = make([]bool, e.concurrency)
+	e.hasNull = make([]bool, e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		e.innerChunk[i] = newFirstChunk(e.innerExecs[i])
+		e.innerList[i] = chunk.NewList(retTypes(e.innerExecs[i]), e.initCap, e.maxChunkSize)
+		e.innerList[i].GetMemTracker().SetLabel(innerListLabel)
+		e.innerList[i].GetMemTracker().AttachTo(e.memTracker)
+	}
+
+	e.freeChkCh = make(chan *chunk.Chunk, e.concurrency)
+	e.resultChkCh = make(chan result, e.concurrency+1) // innerWorkers + outerWorker
+	e.outerRowCh = make(chan outerRow)
+	e.exit = make(chan struct{})
+	for i := 0; i < e.concurrency; i++ {
+		e.freeChkCh <- newFirstChunk(e)
+	}
+	return nil
+}
+
+// Next implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
+	if !e.started {
+		e.wg.Add(1)
+		go e.outerWorker(ctx)
+		for i := 0; i < e.concurrency; i++ {
```

Comment:
This way would use many goroutine and memory for high concurrency.

Reply:
emm, would you have any suggestions?

Reply:
We may need to monitor the consumed resouces, especially considering OOM.
We may put more efforts to rewrite queries to avoiding taking apply.